thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp

```diff
@@ -159,20 +253,92 @@ class CJoinGroup : public CSimpleInterfaceOf<IInterface>
                 }
                 ++cur;
             }
+            while (--numRowArrays);
             ReleaseRoxieRow(rowArrays);
             rowArrays = nullptr;
-            totalRows = 0;
+            totalRhsSz = 0;
+        }
+    }
+    void freeRowsAndStream()
+    {
+        ::Release(rowStream);
+        rowStream = nullptr;
+        freeRows();
+    }
+
+    inline const void *_getNextRhsClear(offset_t &fpos, JoinGroupRhsState &rhsState)
+    {
+        while (rhsState.arr != (rowArrays+numRowArrays)) // end of array marker
+        {
+            if (rhsState.arr->rows)
+            {
+                while (rhsState.pos < rhsState.arr->numRows)
+                {
+                    Row &row = rhsState.arr->rows[rhsState.pos++];
+                    if (row.rhs)
+                    {
+                        fpos = row.fpos;
+                        const void *ret = row.rhs;
+                        row.rhs = nullptr;
+                        return ret;
+                    }
+                }
+            }
+            rhsState.arr++;
+            rhsState.pos = 0;
         }
+        fpos = 0;
+        return nullptr;
     }
+    inline const void *_getFirstRhsClear(offset_t &fpos, JoinGroupRhsState &rhsState)
+    {
+        if (!rowArrays)
+        {
+            fpos = 0;
+            return nullptr;
+        }
+        rhsState.arr = &rowArrays[0];
+        rhsState.pos = 0;
+        return _getNextRhsClear(fpos, rhsState);
+    }
+
+    class CSpillSafeBlock
```

Comment:
Not convinced this is the correct encapsulation.  It is not clear at the point of use that

CriticalBlock b(crit)
and
CSpillSafeBlock cb(this);

are actually protecting using the same critical section.

Reply:
ok, well yes they are, the formed is used inside a CJoinGroup, whereas the latter is used outside and passes a CJoinGroup in.

The code inside the CSpillSafeBlock ctor/dtor that work on the CJoinGroup, could move inside CJoinGroup. I think that would be an improvement.